### ISSUE TITLE
Autoencoding Under Normalization Constraints

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#123]
*ICML 2021*

# TL;DR

- Autoencoders can be used in OOD detection (high recon loss implies OOD)
- Conventional AEs tend to reconstruct even the OOD samples, hence limiting its ability to detect OOD.
- Frame autoencoder as EBM and constrain its ability to reconstruct negative samples via efficient MCMC initialization.

# Previous Works

- MCMC: Contrastive Divergence and Persistent Contrastive Divergence

# Method

## Observation
![image](https://user-images.githubusercontent.com/46308407/133914244-033a43f6-b5dc-4215-b7c2-5513d3a4648b.png)

- Autoencoder trained on MNIST can reconstruct samples from other datasets,

## Energy-Based Models

- Energy term is defined as negative log-likelihood of current density estimate.
![image](https://user-images.githubusercontent.com/46308407/133914274-51353a12-e851-46ac-9785-bbb676de2810.png)
![image](https://user-images.githubusercontent.com/46308407/133914275-33901c04-1fa1-4838-9897-be6c3be07d35.png)

- EBMs learn Maximum Likelihood Parameters by assigning higher density on ***observed samples*** and lower density on ***plausible samples under current density estimation***. 
- The training converges when current density estimation is ***close enough*** to the true data distribution, hence canceling off the below gradient terms.

![image](https://user-images.githubusercontent.com/46308407/133914266-34c313a5-78e0-492b-ae5c-9d51f1b0d660.png)

- Graphical illustration
![image](https://user-images.githubusercontent.com/46308407/133914276-6e132c46-6d98-4dc6-b60c-443dae1ece4c.png)

- Negative samples are usually obtained from MCMC (Langevin Markov Chain) process, which can be computationally expensive.
![image](https://user-images.githubusercontent.com/46308407/133914284-890ab07a-d3d4-4225-b1a3-043e66c593d1.png)

## Normalized Auto-Encoder (NAE)
- Define energy term to be the reconstruction loss.

![image](https://user-images.githubusercontent.com/46308407/133914291-d9767068-bf46-4208-bba7-9500edfd1f96.png)

- Density estimate can be obtained as belows (regarding energy term as the negative log likelihood)
![image](https://user-images.githubusercontent.com/46308407/133914300-5cdeaed0-853b-4a32-b2c5-af44d2551f48.png)

![image](https://user-images.githubusercontent.com/46308407/133914303-4c595f36-d0ce-4c5d-8046-d87df9b10d73.png)

- Optimize in conventional EBM framework.
- That is, minimize reconstruction loss for real sample and maximize it for ***MCMC-generated*** samples.

## On-Manifold Initialization: Efficient MCMC initialization

- With MCMC, we want to sample from our estimated density. (more samples on high density region, vice versa)
- Previous works initialized the chain either from (1) observed data point or (2) previous end point (which started from observed data point as well)
- Each correspond to CD (Contrastive Divergence) and PCD (Persistent Contrastive Divergence)
- However, with well-trained autoencoder, we can assume that decoder outputs are assigned relatively high density (as they are literally ***observed***)
- Hence, initialize the chain from decoder output instead of observed data point.

![image](https://user-images.githubusercontent.com/46308407/133914357-e9e61f40-9dd6-4842-83ee-15247a092a09.png)

![image](https://user-images.githubusercontent.com/46308407/133914355-f0566c49-49be-4622-b12a-38227ba615c4.png)

![image](https://user-images.githubusercontent.com/46308407/133914401-a3358230-044f-462a-a4a5-285b227b34ae.png)

# Experiments

![image](https://user-images.githubusercontent.com/46308407/133914377-c9109a0d-0d0a-4448-91dc-b0b4471d3dbc.png)
![image](https://user-images.githubusercontent.com/46308407/133914379-31884b6d-4f14-4f2d-87e3-fa7177f24a89.png)
![image](https://user-images.githubusercontent.com/46308407/133914387-61c21c68-1230-452f-b8eb-0d63b4f3fc71.png)
![image](https://user-images.githubusercontent.com/46308407/133914411-f940d795-ad18-4bfc-a851-5648ab64d7b6.png)

# Comments

- Didn't know that autoencoders had such practical usage: OOD detection.
- Good to note the definition of OOD: alpha-sublevel set on density function.
- Note that SVHN and CIFAR are ***out-of-support*** as they can be perfectly distinguished by a trained classifier.
